Performer

DrewGalbraith opened the pull request:
This is our long-awaited Performer PR. Congrats and thank you to @RLin8103 and @JacksonSearle for getting us this far! I merged main into Performer yesterday. It was a big commit, but I think I actually did it right. I've never done gone the other direction though, so I'd appreciate a thorough lookover. 

I can train transformer and retnet models still, so that's confirmed tested. I haven't tried a Performer @JacksonSearle, when you do one from the Performer branch for the final paper, would you drop a comment if you have any problems or not? Maybe launch one before you pull the main-->Performer merge and then pull, launch again, and report.

Jackson reduced a lot of code duplication, removing the `models.py` file in favor of seperate files in a new dir. It's pretty slick.

I'm happy to try and answer any other questions, alongside @RLin8103 and @JacksonSearle.

RLin8103 committed:
Add Skeleton Structure to README.md

RLin8103 committed:
Fix Format Issues

RLin8103 committed:
Fix Code Format Issue in Readme

RLin8103 committed:
Add More Titles to the Readme Skeleton

RLin8103 committed:
Add Python Version Requirements to Installation and Setup

RLin8103 committed:
Add Mamba Instruction to Setup

RLin8103 committed:
Fix Citation Format

RLin8103 committed:
Fix Retnet Paper Link and Citation Version

RLin8103 committed:
Update Retnet Paper Link

RLin8103 committed:
Add Preparing the Environment and Execution Workflow Steps

RLin8103 committed:
Add Link to Paper and Clean Up Some Extra Information

RLin8103 committed:
Add Code Instruction for Execution Workflow Section

RLin8103 committed:
Add Paper Link to Results Section and Add Information to Acknowledgements Section

RLin8103 committed:
Fix Formatting

RLin8103 committed:
Add Grid Search Feature Description

RLin8103 committed:
Add Hugging Face Feature Description and Fix Formatting Issues

RLin8103 committed:
Add Lightning Feature Description

RLin8103 committed:
Add Torchscale Feature Description

RLin8103 committed:
Fix Some Wording

RLin8103 committed:
Fix Some Wording of TorchScale Feature Description

DrewGalbraith committed:
Update README.md

Break the link to overleaf all the way

DrewGalbraith committed:
Update README.md

Correct title of paper cited

DrewGalbraith committed:
Update README.md

Punctuation fixes and command unification

DrewGalbraith committed:
Update README.md with nvidia command and correct grid_search parameters

DrewGalbraith committed:
Make README.md a little prettier

DrewGalbraith committed:
Update README.md with correct grid_search parameters

DrewGalbraith committed:
Update README.md

I added the number of unique models generated by our grid search data_run.

DrewGalbraith committed:
Delete unecessary line from nvidia section.

JayOrten committed:
add section on yaml

JayOrten committed:
add setup to execution

DrewGalbraith committed:
Upload a PDF version of Retnet paper draft 

Upload a PDF version of Retnet paper draft to link into read me while we figure out Overleaf non-editable link sharing.

DrewGalbraith committed:
Update README.md with pdf link to paper draft

DrewGalbraith committed:
Update README.md with lite benchmark info

DrewGalbraith committed:
Alphabetize citations in README.md

DrewGalbraith committed:
Restructure acknowledgements in README.md

DrewGalbraith committed:
Clarify benchmarks section of README.md

DrewGalbraith committed:
Update README.md with agnostic mamba environement name

DrewGalbraith committed:
Update README.md titlt to refelct project title

Most popular projects will use the project/paper title as their README.md title. (See BigBench, BBH, ElutherAI LM_eval, etc.)

DrewGalbraith committed:
capitalize mamba placeholder to <YOUR_ENV_NAME>

nprisbrey committed:
Change Wording Of Some Sections

nprisbrey committed:
Fix Relative Links

Commit:
Delete Others Section

Commit:
Change Description For TorchScale RetNet and Transformer

RLin8103 committed:
Change <some_script> to <SOME_SCRIPT>

RLin8103 committed:
Delete Node On Cuda Section

RLin8103 committed:
Change all needed scripts to all of them

RLin8103 committed:
Add cd ./scripts/user_scripts in Execution Workflow

RLin8103 committed:
Delete Citation References And The Section Simple

RLin8103 committed:
Delete Citation References And The Section Simple

RLin8103 committed:
Update README.md

Fix Typo out to our

RLin8103 committed:
Update README.md

Delete Check Python Version Code Instructions.

DrewGalbraith committed:
Fix typo and correct a file path

DrewGalbraith committed:
Add benchamrking instructions

DrewGalbraith committed:
Update README.md with yaml string warning

DrewGalbraith committed:
Remove extraneous text from setup instructions.

DrewGalbraith committed:
Cut general linux tutorials

DrewGalbraith committed:
General rewording and editing

DrewGalbraith committed:
Cut more extraeneous instructions

nprisbrey committed:
Enumerate Installation and Setup Steps

nprisbrey committed:
Reword Environment Setup Instructions

nprisbrey committed:
Tweak Relative Paths

nprisbrey committed:
Formatting Tweaks For README

RLin8103 committed:
Update README.md

Resolve Code Comments Format Issues In Benchmark

RLin8103 committed:
Delete weights directory

RLin8103 committed:
Delete scripts/job.sh

RLin8103 committed:
Delete Period In Comment

nprisbrey committed:
Fix Path Error In README

nprisbrey committed:
Fix Typos In README

DrewGalbraith committed:
Delete 301R_Retnet_Paper.pdf

DrewGalbraith committed:
Enhance writing

DrewGalbraith committed:
edit writing

DrewGalbraith committed:
Update README.md with train_model.sh

DrewGalbraith committed:
Elimante links to deleted pdf

DrewGalbraith committed:
Add benchmark file path

DrewGalbraith committed:
Remove paper title from README.md

DrewGalbraith committed:
Remove paper title from README.md

RLin8103 committed:
Add Per Model Architecture After 72 Combination

RLin8103 committed:
Change Acknowledgement And Fix Minor Grammatic Issue

RLin8103 committed:
Add Note In Script Execution To Highlight

RLin8103 committed:
Change Note To Tip

RLin8103 committed:
Add Note To Citations And Change Wording

RLin8103 committed:
Create Performer Python File From Decoder

RLin8103 committed:
Create Performer Python File From Decoder

RLin8103 committed:
Add New File With Fast Attention, Causal Linear Attention and Other Necessary Functions

RLin8103 committed:
Add New File With Fast Attention, Causal Linear Attention and Other Necessary Functions

RLin8103 committed:
Add Attention Folder and Functions, and Change Self Attention and Cross Attention in Decoder Class

RLin8103 committed:
Add Attention Folder and Functions, and Change Self Attention and Cross Attention in Decoder Class

RLin8103 committed:
Add Imports and Helper Functions to Performer Attention Class

RLin8103 committed:
Add Imports and Helper Functions to Performer Attention Class

RLin8103 committed:
Add Local Attention Class

RLin8103 committed:
Add Local Attention Class

RLin8103 committed:
Add Performer to model.py

RLin8103 committed:
Add Performer to model.py

RLin8103 committed:
Add Performer to train_model.py

RLin8103 committed:
Add Performer to train_model.py

RLin8103 committed:
Add Performer Configs to config.py

RLin8103 committed:
Add Performer Configs to config.py

RLin8103 committed:
Merge implement_performer.py with performer.py

RLin8103 committed:
Merge implement_performer.py with performer.py

RLin8103 committed:
Delete torchscale/architecture/implement_performer.py

RLin8103 committed:
Delete torchscale/architecture/implement_performer.py

RLin8103 committed:
Add Comments to perfomer.py

RLin8103 committed:
Add Comments to perfomer.py

RLin8103 committed:
Add Necessary Imports and Classes for Performer

RLin8103 committed:
Add Necessary Imports and Classes for Performer

RLin8103 committed:
Merge branch 'performer' of https://github.com/DRAGNLabs/301r_retnet into performer

RLin8103 committed:
Make axial_positional_embedding A Python File

RLin8103 committed:
Make axial_positional_embedding A Python File

RLin8103 committed:
Delete torchscale/component/performer/axial_positional_embedding

RLin8103 committed:
Delete torchscale/component/performer/axial_positional_embedding

RLin8103 committed:
Add More Necessary Classes

RLin8103 committed:
Add More Necessary Classes

RLin8103 committed:
Merge branch 'performer' of https://github.com/DRAGNLabs/301r_retnet into performer

RLin8103 committed:
Add Comments to Performer and PerformerLM Class in attention.py

RLin8103 committed:
Add Comments to Performer and PerformerLM Class in attention.py

RLin8103 committed:
Add Performer Positional Embedding initialization and Calculation in PerformerDecoder Class

RLin8103 committed:
Add Performer Positional Embedding initialization and Calculation in PerformerDecoder Class

RLin8103 committed:
Change dropout Back to dropout_module

RLin8103 committed:
Change dropout Back to dropout_module

RLin8103 committed:
Fix Import Names and Class Names

RLin8103 committed:
Fix Import Names and Class Names

RLin8103 committed:
Fix Performer's Config to Align with Decoder

RLin8103 committed:
Fix Performer's Config to Align with Decoder

RLin8103 committed:
Add Performer Configs to Yaml file and Add Embedding Default Parameters to PerformerDecoder

RLin8103 committed:
Add Performer Configs to Yaml file and Add Embedding Default Parameters to PerformerDecoder

RLin8103 committed:
Add Performer Config Parameters' Value

RLin8103 committed:
Add Performer Config Parameters' Value

KimballJardine committed:
Switch args.decoder_layers to args.layers

KimballJardine committed:
Switch args.decoder_layers to args.layers

RLin8103 committed:
Delete Encoder and EncoderDecoder Config

RLin8103 committed:
Delete Encoder and EncoderDecoder Config

RLin8103 committed:
Add Performer to run_eval.py:

RLin8103 committed:
Add Performer to run_eval.py:

KimballJardine committed:
Undo args.layer changes

KimballJardine committed:
Undo args.layer changes

RLin8103 committed:
Change PerformerConfig in model.py

RLin8103 committed:
Deleted Unused Performer Configuration in config.py

KimballJardine committed:
Merge remote-tracking branch 'origin' into performer

KimballJardine committed:
Test commit.

KimballJardine committed:
Merge remote-tracking branch 'origin/performer' into performer

KimballJardine committed:
Remove duplicate field in config.yaml.

KimballJardine committed:
remove test message

RLin8103 committed:
Add MOE Configs to PerformerConfig

RLin8103 committed:
Fix File Paths For Performer Component

RLin8103 committed:
Add Performer Specific Config to PerformerModel

RLin8103 committed:
Delete Unused Performer Config in template.config.yaml

RLin8103 committed:
Add performer as a Choice in model_type comment in the YAML file

RLin8103 committed:
Update requirements.txt to Align with main

RLin8103 committed:
Add Comma after config.vocab_size in models.py

RLin8103 committed:
Fix max_seq_len Configs

RLin8103 committed:
Fix Performer attention.py Imports

RLin8103 committed:
Add deepnorm and subln to PerformerConfig

RLin8103 committed:
Add max_seq_len to PerformerHF

RLin8103 committed:
Add activation_dropout and no_scale_embedding

JacksonSearle committed:
adds necessary PerformerDecoder arguments

JacksonSearle committed:
Merge branch 'performer' of https://github.com/DRAGNLabs/301r_retnet into performer

RLin8103 committed:
Add no_output_layer to PerformerConfig

RLin8103 committed:
Add share_decoder_input_output_embed

RLin8103 committed:
Add layernorm_embedding to PerformerConfig

RLin8103 committed:
Add layernorm_eps to PerformerConfig

RLin8103 committed:
Change activation_function to activation_fn

RLin8103 committed:
Add Fairscale Configs to PerformerConfig

RLin8103 committed:
Add rel_pos_buckets to PerformerConfig

RLin8103 committed:
Add bert_init and multiway

RLin8103 committed:
Current Version of Performer Forward Embedding Logic

JacksonSearle committed:
debugging print messages

JacksonSearle committed:
Merge branch 'main' into performer

Conflicts:
	configs/template_config.yaml
	requirements.txt

JacksonSearle committed:
no quotes

JacksonSearle committed:
uses correct config variable in performer

JacksonSearle committed:
replacing print statements

JacksonSearle committed:
performerhf fix to be more like other hf classes

JacksonSearle committed:
model inheritance

JacksonSearle committed:
updated imports with architecture folder rename

JacksonSearle committed:
removed self.config = config, this is already in the parent class

JacksonSearle committed:
renamed files for naming clarity

JacksonSearle committed:
fixing build_self_attention args to be from args

RLin8103 committed:
Change CrossAttention Arguments to be from args

RLin8103 committed:
Fix Missing x Argument

JacksonSearle committed:
sets all embed dimensions to be the same

JacksonSearle committed:
Merge branch 'performer' of https://github.com/DRAGNLabs/301r_retnet into performer

JacksonSearle committed:
removes attn variable used in encoder

JacksonSearle committed:
ignores unused parameters

JacksonSearle committed:
delete extranious comments

JacksonSearle committed:
add new performer arguments

DrewGalbraith committed:
Merge branch 'main' into performer

DrewGalbraith committed:
Add LongNet to class

nprisbrey commented on configs/template_config.yaml:
```diff
@@ -172,14 +170,40 @@ heads: 4
 layers: 2
 # Model Type (str): Name of model architecture to train
-model_type: "retnet" # Choices: "retnet", "transformer", "longnet"
+model_type: retnet # Choices: "retnet", "transformer", "performer"
```
Why are we removing "longnet"?

RLin8103 replied:
This issue is fixed.

nprisbrey commented on configs/template_config.yaml:
It doesn't make sense that we're using the C4 dataset but saving tokenizers and other things to "wikitext" folders. This needs to be brought into agreement with each other.

RLin8103 replied:
This issue is fixed

nprisbrey commented on configs/template_config.yaml:
```diff
@@ -172,14 +170,40 @@ heads: 4
+no_projection: false
+# Use bias in QKV projection within the attention mechanism. Determines if bias terms are added to queries, keys, and values.
+qkv_bias: false
+# Use bias in the attention output projection. Determines if a bias term is added to the output of the attention layer.
```
It would be nice if these comments followed the `English Name (type): Description` format precedence of the previous parameters to keep things consistent.

RLin8103 replied:
This issue is fixed.

nprisbrey commented on src/generate.py:
```diff
@@ -18,11 +20,11 @@ def prepare_model_device(config: Struct):
     if config.model_type.lower() == "longnet":
         model = LongNetModel(config)
```
Why doesn't this get changed to `LongNetLightning` as well?

RLin8103 replied:
This issus is fixed.

nprisbrey commented on src/inference.py:
```diff
@@ -25,9 +27,11 @@ def inference(config: Struct):
     if config.model_type.lower() == "longnet":
         model = LongNetModel(config)
```
Why doesn't this get changed to `LongNetLightning` as well?

RLin8103 replied:
This issus is fixed.

nprisbrey commented on src/train_model.py:
```diff
@@ -9,13 +9,18 @@
+from torchinfo import summary as model_summary
+from architecture.lightning.performer_lightning import PerformerLightning
+from architecture.lightning.retnet_lightning import RetNetLightning
+from architecture.lightning.transformer_lightning import TransformerLightning
```
Why are we allowed to leave out `LongNetLightning` from this import list?

RLin8103 replied:
This issus is fixed.

nprisbrey commented on src/train_model.py:
```diff
@@ -81,9 +86,11 @@ def train_model(config: Struct):
     if config.model_type.lower() == "longnet":
         model = LongNetModel(config)
```
Same as previous comments about `LongNetLightning`.

RLin8103 replied:
This issus is fixed.

nprisbrey reviewed (changes requested):
This is a very surface level review. Let me know your thoughts on these small changes and then I'll make another in-depth review. I'll be following the README for the in-depth review, so it would be nice if that was updated for Performer models as well (I haven't checked, though I don't see the README file as one of the updated files in this Pull Request).

DrewGalbraith committed:
Implement reccommendation to change precision

@JayOrten reccommended changing precision from `fp16` (or `bf16`, depending on where in our repo one is) to simply `16`. This is according to the latest TorchLightning docs. TorchLightning, btw, is the package using this precision argument.

RLin8103 committed:
Add longnet as a model_type in template_config.yaml

RLin8103 committed:
Change Performer Config to Follow English Name (type): Description Format

RLin8103 committed:
Add LongNetLightning to Necessary Files

RLin8103 committed:
Change Tokenizer and Dataset Paths to C4 in Config
